beacon-chain/rpc/eth/validator/validator.go

```diff
@@ -966,6 +967,70 @@ func (vs *Server) SubmitContributionAndProofs(ctx context.Context, req *ethpbv2.
 	return &empty.Empty{}, nil
 }
 
+// GetLiveness requests the beacon node to indicate if a validator has been observed to be live in a given epoch.
+// The beacon node might detect liveness by observing messages from the validator on the network,
+// in the beacon chain, from its API or from any other source.
+// A beacon node SHOULD support the current and previous epoch, however it MAY support earlier epoch.
+// It is important to note that the values returned by the beacon node are not canonical;
+// they are best-effort and based upon a subjective view of the network.
+// A beacon node that was recently started or suffered a network partition may indicate that a validator is not live when it actually is.
+func (vs *Server) GetLiveness(ctx context.Context, req *ethpbv2.GetLivenessRequest) (*ethpbv2.GetLivenessResponse, error) {
+	ctx, span := trace.StartSpan(ctx, "validator.GetLiveness")
+	defer span.End()
+
+	var participation []byte
+
+	// First we check if the requested epoch is the current epoch.
+	// If it is, then we won't be able to fetch the state at the end of the epoch.
+	// In that case we get participation info from the head state.
+	// We can also use the head state to get participation info for the previous epoch.
+	headSt, err := vs.HeadFetcher.HeadState(ctx)
+	if err != nil {
+		return nil, status.Error(codes.Internal, "Could not get head state")
+	}
+	currEpoch := slots.ToEpoch(headSt.Slot())
+	if req.Epoch > currEpoch {
+		return nil, status.Errorf(codes.InvalidArgument, "Requested epoch cannot be in the future")
+	}
+
+	if req.Epoch == currEpoch {
+		participation, err = headSt.CurrentEpochParticipation()
+		if err != nil {
+			return nil, errParticipation
+		}
+	} else if req.Epoch == currEpoch-1 {
+		participation, err = headSt.PreviousEpochParticipation()
+		if err != nil {
+			return nil, errParticipation
+		}
+	} else {
+		epochEnd, err := slots.EpochEnd(req.Epoch)
+		if err != nil {
+			return nil, status.Error(codes.Internal, "Could not get requested epoch's end slot")
+		}
+		st, err := vs.StateFetcher.StateBySlot(ctx, epochEnd)
+		if err != nil {
+			return nil, status.Error(codes.Internal, "Could not get slot for requested epoch")
+		}
+		participation, err = st.CurrentEpochParticipation()
+		if err != nil {
+			return nil, errParticipation
+		}
+	}
+
+	resp := &ethpbv2.GetLivenessResponse{
+		Data: make([]*ethpbv2.GetLivenessResponse_Liveness, len(req.Index)),
+	}
+	for i, vi := range req.Index {
+		resp.Data[i] = &ethpbv2.GetLivenessResponse_Liveness{
+			Index:  vi,
+			IsLive: participation[vi] != 0,
```

Comment:
Could this panic if vi is outside of the range of `len(participation)` ?

Reply:
If so, add a regression test please.

Reply:
Yes, if the validator index is invalid. I added a defensive check.